### PR TITLE
Improve signer and combiner domain.test.ts

### PR DESCRIPTION
**Pull request by eelanagaraj**
### Description
- uses `ThresholdPoprfClient` to evaluate partial signatures returned by `DOMAIN_SIGN` signer endpoints
  - ⚠️❓ currently just checks against a hard-coded evaluation result, not sure if there's a better way of doing this
- fix test cases with nonces (previously was failing due to incorrect authorization, not nonce being invalid) (in both signer and combiner tests)
  - add case with nonce >0 being valid (this is allowed by the current domain signer logic -- if we don't want this to be the case, we need to update the logic)
  - fix case with repeated request + nonce being incremented (valid) -- keeps the request the same except for the nonce (including blinded message, sessionID being the same -- wasn't the case before)
  - fix nonce failure case -- requires repeat request + nonce not being updated
- modifY `DomainRestrictedSignatureResponse` slightly (allows status field to not be present), moved TODO from tests -> `sendFailure`
- passthrough 401 errors from the signer as nonce errors in the combiner (since auth is already checked on the combiner side, it seems more likely that 401s are caused by nonce issues and it's pretty confusing to only get the missing signatures error) -- happy to change this back though!

Will merge after the POPRF changes are approved since this is based off of those changes!

### CODE REVIEW

File: packages/phone-number-privacy/signer/src/domain/endpoints/sign/io.ts [hunk 1]
```diff
@@ -77,7 +77,7 @@ export class DomainSignIO extends IO<DomainRestrictedSignatureRequest> {
     error: ErrorType,
     status: number,
     response: Response<DomainRestrictedSignatureResponseFailure>,
-    domainState?: DomainState
+    domainState?: DomainState // TODO: is status ever provided on failure?
```
Comment (alecps) [inline packages/phone-number-privacy/signer/src/domain/endpoints/sign/io.ts:80]: If I remember correctly, the domain state should be provided on failure when the requests fails due to the user being out of quota or the domain being disabled. That is, the domain state should be provided whenever it can explain why the failure happened

File: packages/phone-number-privacy/combiner/src/common/sign.ts [hunk 2]
```diff
@@ -71,10 +71,19 @@ export abstract class SignAction<R extends OdisSignatureRequest> extends Combine
...
     const majorityErrorCode = session.getMajorityErrorCode()
-    if (majorityErrorCode === 403 || majorityErrorCode === 429) {
-      error = WarningMessage.EXCEEDED_QUOTA
+    switch (majorityErrorCode) {
+      case 403:
+      case 429:
+        error = WarningMessage.EXCEEDED_QUOTA
+        break
+      case 401:
+        // Authentication is checked in the combiner, but invalid nonces are passed through
```
Comment (alecps) [inline packages/phone-number-privacy/combiner/src/common/sign.ts]: I apologize for setting us up with poor factoring here, but it looks like maybe we should consider making handleMissingSignatures abstract and moving the implementation(s) into the domain/sign and pnp/sign files. My thinking is that we use different status codes in each context. For example, there's no notion of a 'nonce' in the pnp use case. Wdyt?

File: packages/phone-number-privacy/signer/test/integration/domain.test.ts [hunk 3]
```diff
@@ -476,16 +486,21 @@ describe('domainService', () => {
...
         },
       })
+      const eval1 = thresholdPoprfClient.unblindPartialResponse(
+        Buffer.from(res1.body.signature, 'base64')
+      )
+      expect(eval1.toString('base64')).toEqual(expectedEval)
 
       // submit identical request with nonce set to 1
-      const req2 = (await signatureRequest(undefined, 1))[0]
+      req1.options.nonce = defined(1)
```
Comment (alecps) [inline packages/phone-number-privacy/signer/test/integration/domain.test.ts]: nit: consider defining req2

File: packages/phone-number-privacy/signer/test/integration/domain.test.ts [hunk 4]
```diff
@@ -476,16 +486,21 @@ describe('domainService', () => {
...
+      req1.options.signature = noString
```
Comment (alecps) [inline packages/phone-number-privacy/signer/test/integration/domain.test.ts]: Why do we have to reset to noString before defining a new value for the signature?

File: packages/phone-number-privacy/signer/test/integration/domain.test.ts (hunk 3 above)
Comment (alecps) [inline packages/phone-number-privacy/signer/test/integration/domain.test.ts]: Oh weird...I guess this should've been failing with 401 before (or if we remove the nonce update)? Do you know why it was passing? I'm guessing you've explained this and I'm missing something 😅

File: packages/phone-number-privacy/signer/test/integration/domain.test.ts [hunk 5]
```diff
@@ -498,10 +513,35 @@ describe('domainService', () => {
...
         },
       })
+      const eval2 = thresholdPoprfClient.unblindPartialResponse(
+        Buffer.from(res2.body.signature, 'base64')
+      )
+      expect(eval2).toEqual(eval1)
+    })
+
+    it('Should respond with 200 if nonce > domainState', async () => {
+      const [req, thresholdPoprfClient] = await signatureRequest(undefined, 1)
```
Comment (alecps) [inline packages/phone-number-privacy/signer/test/integration/domain.test.ts]: Nit: consider submitting this request with nonce == 2 to make it extra clear and also so that we're testing whether counter just gets incremented vs. set all the way up to the nonce value. (i.e. will counter be returned as 1 or 2 if we submit with nonce 2)

File: packages/phone-number-privacy/signer/test/integration/domain.test.ts [hunk 6]
```diff
@@ -607,26 +646,42 @@ describe('domainService', () => {
...
         error: WarningMessage.UNAUTHENTICATED_USER,
       })
     })
 
-    it('Should respond with 401 on invalid nonce', async () => {
-      const [badRequest, _] = await signatureRequest()
-      badRequest.options.nonce = defined(1)
-
-      const res = await request(app).post(SignerEndpoint.DOMAIN_SIGN).send(badRequest)
+    it('Should respond 401 on invalid nonce', async () => {
```
Comment (alecps) [inline packages/phone-number-privacy/signer/test/integration/domain.test.ts:657]: Nice catch adding this!!

**Review by alecps — APPROVED** [review]
Looks great!! Just a few comments / questions. Thanks for going the extra mile to identify these fixes 🙏🙏

File: packages/phone-number-privacy/combiner/src/common/sign.ts (hunk 2 above)
Comment (alecps) [reply packages/phone-number-privacy/combiner/src/common/sign.ts]: Feel free to add a TODO for this and we can get to it later

Comment (eelanagaraj) [reply packages/phone-number-privacy/combiner/src/common/sign.ts]: Edit: modified it slightly, we can choose a different structure altogether but lmk what you think of this one @alecps !

Was also wondering if it makes sense to define a different error code for nonce errors? To be fair, I couldn't think of what classical error would fit better, but it could be really nice for error handling to know the difference between signature issues + nonce issues?

File: packages/phone-number-privacy/signer/test/integration/domain.test.ts (hunk 3 above)
Comment (eelanagaraj) [reply packages/phone-number-privacy/signer/test/integration/domain.test.ts]: I'll change the name from `req1` -> `req`, though I intentionally want the request to be the exact same except for the nonce/signature. Otherwise, I think it creates room for the issue before which is that there's another difference in the requests (not just the nonce) that is sneakily causing success. Wdyt?

Comment (eelanagaraj) [reply packages/phone-number-privacy/signer/test/integration/domain.test.ts]: Yeah! So before, the test case for the repeated requests was not only testing the nonce change -- we were creating a separate request with an entirely new blinded message + session, so while this was a repeated request for a single account, it didn't really test the exact same request overall (since the message, session, and nonce were all different). The previous nonce failure test case (below in the diff, I can tag you down there!) was failing because of the signature over the request not being correct -- we updated the nonce but didn't delete the old signature (the `NoString` part), re-sign the message, and include this signature in the new request.

File: packages/phone-number-privacy/signer/test/integration/domain.test.ts (hunk 4 above)
Comment (eelanagaraj) [reply packages/phone-number-privacy/signer/test/integration/domain.test.ts]: Similar to the above but slightly more detail: if we don't reset it to `noString`, the signature is over the request (including the old signature), which then doesn't match (since we don't end up sending the old sig in the request). The EIP712 data structure that gets signed (and that the server-side reconstructs) has the options.signature field (i.e. can't delete it outright -- we will get an error since the field needs to be defined even if it doesn't have a value), but it also expects the field to be empty hence the `NoString`. Hope that makes sense!!

File: packages/phone-number-privacy/signer/test/integration/domain.test.ts (hunk 5 above)
Comment (eelanagaraj) [reply packages/phone-number-privacy/signer/test/integration/domain.test.ts]: Good call! Changed + added a comment; counter gets incremented, **not** set to the value of the nonce. Is this the desired behavior? (Seems like someone could replay messages with a nonce higher than the counter, but maybe that's fine/an accepted tradeoff here?)

**Comment by alecps** [comment]
Nice! I like what you did with errorCodeToError

File: packages/phone-number-privacy/signer/src/domain/endpoints/sign/io.ts (hunk 1 above)
Comment (eelanagaraj) [reply packages/phone-number-privacy/signer/src/domain/endpoints/sign/io.ts:80]: forgot to respond to this earlier! This is just moving a TODO from the tests -> location of the change; I agree with you though I don't know that there are any instances of us actually sending domain state. Leaving the TODO for now since I didn't look into it in depth as part of this PR!

File: packages/phone-number-privacy/signer/test/integration/domain.test.ts (hunk 5 above)
Comment (alecps) [reply packages/phone-number-privacy/signer/test/integration/domain.test.ts]: This is a somewhat subtle point that Victor and I discussed. The only reason the client should generate a request with a nonce higher than counter+1 is if it needs to in order to reach a threshold of signers (i.e. if the signers have different views of what counter the client is on). In this case, we want to preserve the user's quota on each signer in case the discrepancy in signer states changes direction again. A replay attack here could only deplete the user's quota up to the point that setting counter == nonce would anyway, and the replayed requests don't pose any additional security threat beyond depleting the user's quota unless a threshold of signers can be reached. It's a pretty subtle edge case and in general either behavior would be fine here. This is the expected behavior we landed on though IIRC. Let me know if you have any thoughts on this!

File: packages/phone-number-privacy/signer/test/integration/domain.test.ts (hunk 3 above)
Comment (alecps) [reply packages/phone-number-privacy/signer/test/integration/domain.test.ts]: Ahh makes sense! Great catch

File: packages/phone-number-privacy/signer/test/integration/domain.test.ts (hunk 5 above)
Comment (eelanagaraj) [reply packages/phone-number-privacy/signer/test/integration/domain.test.ts]: Sounds good, thanks for the context!

File: packages/phone-number-privacy/signer/test/integration/domain.test.ts (hunk 4 above)
Comment (alecps) [reply packages/phone-number-privacy/signer/test/integration/domain.test.ts]: Yes that makes perfect sense! Thanks!!

File: packages/phone-number-privacy/combiner/src/common/sign.ts (hunk 2 above)
Comment (alecps) [reply packages/phone-number-privacy/combiner/src/common/sign.ts]: Hmm I'm not against this but it seems like right now we're able to disambiguate based on the error message, right?

Comment (eelanagaraj) [reply packages/phone-number-privacy/combiner/src/common/sign.ts]: Yes (enough) since we also check signatures at the combiner level too (i.e. we duplicate those checks). That said, if there is a case where the combiner incorrectly forwards a request with bad auth, we can't really tell the difference right now

Comment (alecps) [reply packages/phone-number-privacy/combiner/src/common/sign.ts]: Gotcha! I created this ticket to track https://github.com/orgs/celo-org/projects/164/views/28
